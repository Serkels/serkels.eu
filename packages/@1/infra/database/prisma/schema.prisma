//

//
//
//

//
//
//

datasource db {
  provider = "postgres"
  url      = env("DATABASE_URL")
  // provider = "sqlite"
  // url      = "file:./dev.db"
}

generator client {
  provider = "prisma-client-js"
}

//
//
//

//
//
//

model SignupPayload {
  role    ProfileRole
  name    String
  email   String              @unique
  tokens  VerificationToken[]
  context Json
}

model Profile {
  id        String      @id @default(cuid())
  bio       String?
  bookmarks Bookmark[]
  contacts  Profile[]   @relation("Contacts")
  following Profile[]   @relation("ProfileFollows")
  image     String
  name      String
  role      ProfileRole
  user      User        @relation(fields: [user_id], references: [id], onDelete: Cascade)
  user_id   String      @unique @map("userId")
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  partner                Partner?
  studient               Studient?
  written_messages       Message[] @relation("WrittenMessages")
  followed_by            Profile[] @relation("ProfileFollows")
  participants_in_thread Thread[]  @relation("ParticipantsInThread")
  in_contact_with        Profile[] @relation("Contacts")
}

model Studient {
  id             String     @id @default(cuid())
  citizenship    String?
  city           String?
  field_of_study String?
  interest       Category[] @relation("CategorizedStudientInterest")
  profile        Profile    @relation(fields: [profile_id], references: [id], onDelete: Cascade)
  profile_id     String     @unique
  university     String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  asked_questions      Question[]       @relation("AskedQuestions")
  exchange_threads     ExchangeThread[]
  inbox_threads        InboxThread[]
  participants_in_deal Deal[]
  proposed_exchanges   Exchange[]       @relation("ProposedExchanges")
}

model Partner {
  id            String        @id @default(cuid())
  city          String?
  link          String
  opportunities Opportunity[]
  profile       Profile       @relation(fields: [profile_id], references: [id], onDelete: Cascade)
  profile_id    String        @unique
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
}

//

enum ProfileRole {
  ADMIN
  PARTNER
  STUDIENT
}

//
//
//

//
//
//

model Bookmark {
  id             String       @id @default(cuid())
  exchange       Exchange?    @relation("BookmarkedExchanges", fields: [exchange_id], references: [id])
  exchange_id    String?
  opportunity    Opportunity? @relation("BookmarkedOpportunities", fields: [opportunity_id], references: [id])
  opportunity_id String?
  owner          Profile      @relation(fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id       String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  @@unique(fields: [owner_id, exchange_id], name: "bookmarked_exchange")
  @@unique(fields: [owner_id, opportunity_id], name: "bookmarked_opportunity")
}

//
//
//

//
//
//

model Category {
  id      String           @id @default(cuid())
  context CategoryContext? @default(EXCHANGE)
  rank    Int?
  name    String
  slug    String
  //

  categorized_exchanges         Exchange[]    @relation("CategorizedExchanges")
  categorized_opportunities     Opportunity[] @relation("CategorizedOpportunities")
  categorized_questions         Question[]    @relation("CategorizedQuestions")
  categorized_return_exchanges  Exchange[]    @relation("CategorizedReturnExchanges")
  categorized_studient_interest Studient[]    @relation("CategorizedStudientInterest")
  //

  @@unique(fields: [slug, context], name: "slug_in_context")
}

enum CategoryContext {
  EXCHANGE
  OPPORTUNITY
  FORUM
}

//
//
//

//
//
//

model Exchange {
  id          String       @id @default(cuid())
  category    Category     @relation("CategorizedExchanges", fields: [category_id], references: [id])
  category_id String
  deals       Deal[]
  description String
  expiry_date DateTime     @default(now())
  is_active   Boolean
  is_online   Boolean
  location    String?
  owner       Studient     @relation("ProposedExchanges", fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id    String
  places      Int
  return      Category?    @relation("CategorizedReturnExchanges", fields: [return_id], references: [id])
  return_id   String?
  title       String
  type        ExchangeType
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  bookmarked_exchanges Bookmark[] @relation("BookmarkedExchanges")
}

enum ExchangeType {
  PROPOSAL
  RESEARCH
}

model Deal {
  id             String               @id @default(cuid())
  status         ExchangeThreadStatus @default(IDLE)
  parent         Exchange             @relation(fields: [parent_id], references: [id], onDelete: Cascade)
  parent_id      String
  participant    Studient             @relation(fields: [participant_id], references: [id], onDelete: Cascade)
  participant_id String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  exchange_threads ExchangeThread[]
  //

  @@unique(name: "participant_per_exchange", fields: [parent_id, participant_id])
}

model ExchangeThread {
  id             String   @id @default(cuid())
  last_seen_date DateTime @default(now())
  owner          Studient @relation(fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id       String
  deal           Deal     @relation(fields: [deal_id], references: [id], onDelete: Cascade)
  deal_id        String
  thread         Thread   @relation(fields: [thread_id], references: [id], onDelete: Cascade)
  thread_id      String

  @@unique([owner_id, deal_id])
}

enum ExchangeThreadStatus {
  IDLE
  DENIED
  APPROVED_BY_THE_ORGANIZER
  APPROVED
}

//
//
//

//
//
//

model Opportunity {
  id          String   @id @default(cuid())
  category    Category @relation("CategorizedOpportunities", fields: [category_id], references: [id])
  category_id String
  cover       String
  description String
  expiry_date DateTime
  link        String
  location    String?
  owner       Partner  @relation(fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id    String
  slug        String   @unique
  title       String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  bookmarked_opportunities Bookmark[] @relation("BookmarkedOpportunities")
}

//
//
//

//
//
//

model Question {
  id          String   @id @default(cuid())
  category    Category @relation("CategorizedQuestions", fields: [category_id], references: [id])
  category_id String
  owner       Studient @relation("AskedQuestions", fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id    String
  title       String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
}

//
//
//

//
//
//

model InboxThread {
  id             String   @id @default(cuid())
  is_archived    Boolean  @default(false)
  last_seen_date DateTime @default(now())
  owner          Studient @relation(fields: [owner_id], references: [id], onDelete: Cascade)
  owner_id       String
  thread         Thread   @relation(fields: [thread_id], references: [id], onDelete: Cascade)
  thread_id      String

  @@unique([owner_id, thread_id])
}

model Thread {
  id           String    @id @default(cuid())
  messages     Message[]
  participants Profile[] @relation("ParticipantsInThread")
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
  //

  inbox_threads    InboxThread[]
  exchange_threads ExchangeThread[]
}

model Message {
  id        String  @id @default(cuid())
  author    Profile @relation("WrittenMessages", fields: [author_id], references: [id], onDelete: Cascade)
  author_id String
  content   String
  thread    Thread  @relation(fields: [thread_id], references: [id])
  thread_id String
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())

  @@unique(name: "unique_date_in_thread", fields: [thread_id, created_at])
}

//
//
//

//
//
//

model PasswordlessToken {
  id         String    @id @default(uuid())
  body       String    @unique
  email      String
  login_date DateTime?
  is_active  Boolean   @default(true)
  //

  created_at DateTime @default(now())
  updated_at DateTime @default(now())
}

//

model Account {
  id                String  @id @default(cuid())
  userId            String
  type              String
  provider          String
  providerAccountId String
  refresh_token     String? @db.Text
  access_token      String? @db.Text
  expires_at        Int?
  token_type        String?
  scope             String?
  id_token          String? @db.Text
  session_state     String?

  user User @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model Session {
  id           String   @id @default(cuid())
  sessionToken String   @unique
  userId       String
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  accounts      Account[]
  sessions      Session[]

  //

  profile Profile?
}

model VerificationToken {
  identifier String
  token      String   @unique
  expires    DateTime

  payload       SignupPayload? @relation(fields: [payload_email], references: [email], onDelete: Cascade)
  payload_email String?

  @@unique([identifier, token])
}
