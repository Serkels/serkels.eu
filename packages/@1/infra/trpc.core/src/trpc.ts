//

import { initTRPC } from "@trpc/server";
import SuperJSON from "superjson";
import { ZodError } from "zod";
import type { Context } from "./context";

//

export const { router, middleware, mergeRouters, procedure } = initTRPC
  .context<Context>()
  .create({
    transformer: SuperJSON,
    errorFormatter({ shape, error }) {
      return {
        ...shape,
        data: {
          ...shape.data,
          zodError:
            error.code === "BAD_REQUEST" && error.cause instanceof ZodError
              ? error.cause.flatten()
              : null,
        },
      };
    },
  });

// const isAuthed = middleware(({ next, ctx }) => {
//   const user = ctx.session?.user;

//   if (!user?.name) {
//     throw new TRPCError({ code: 'UNAUTHORIZED' });
//   }

//   return next({
//     ctx: {
//       user: {
//         ...user,
//         name: user.name,
//       },
//     },
//   });
// });

// /**
//  * Protected base procedure
//  */
// export const authedProcedure = t.procedure.use(isAuthed);
