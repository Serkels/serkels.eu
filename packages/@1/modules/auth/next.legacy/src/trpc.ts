//

import type { Router } from "@1.infra/trpc";
import { NEXTAUTH_TRPCENV } from "@douglasduteil/nextauth...trpc.prisma/config";
import { create_nextauth_header } from "@douglasduteil/nextauth...trpc.prisma/jwt";
import { createTRPCProxyClient, httpLink, loggerLink } from "@trpc/client";
import type { JWT } from "next-auth/jwt";
import SuperJSON from "superjson";
import { z } from "zod";

//

const ENV = z
  .object({ API_URL: z.string().url() })
  .merge(NEXTAUTH_TRPCENV)
  .parse(process.env);

//

export const trpc = createTRPCProxyClient<Router>({
  links: [
    loggerLink({
      enabled: (opts) =>
        process.env.NODE_ENV === "development" ||
        (opts.direction === "down" && opts.result instanceof Error),
    }),
    httpLink({
      url: `${ENV.API_URL}/trpc`,
      headers: async () => {
        const nexaut_header = await create_nextauth_header({
          secret: ENV.NEXTAUTH_SECRET,
          token: {
            from: "@1.modules/auth.next.legacy",
            profile: { id: "SSR", image: "", name: "", role: "ADMIN", bio: "" },
          } satisfies JWT,
          maxAge: 60,
        });
        return nexaut_header;
      },
    }),
  ],
  transformer: SuperJSON,
});
