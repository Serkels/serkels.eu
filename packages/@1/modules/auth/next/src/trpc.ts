//

import type { Router } from "@1.infra/trpc";
import { NEXTAUTH_TRPCENV } from "@douglasduteil/nextauth...trpc.prisma/config";
import { create_nexauth_header } from "@douglasduteil/nextauth...trpc.prisma/jwt";
import { createTRPCProxyClient, httpLink, loggerLink } from "@trpc/client";
import type { JWT } from "next-auth/jwt";
import SuperJSON from "superjson";
import { z } from "zod";

//

const ENV = z
  .object({ API_URL: z.string().url() })
  .merge(NEXTAUTH_TRPCENV)
  .parse(process.env);

//

export const trpc = createTRPCProxyClient<Router>({
  links: [
    loggerLink({
      enabled: (opts) =>
        process.env.NODE_ENV === "development" ||
        (opts.direction === "down" && opts.result instanceof Error),
    }),
    httpLink({
      url: `${ENV.API_URL}/trpc`,
      headers: async ({}) => {
        const nexaut_header = await create_nexauth_header({
          salt: "",
          secret: ENV.AUTH_SECRET,
          token: {
            from: "@1.modules/auth.next",
            profile: { id: "SSR", image: "", name: "", role: "ADMIN" },
          } satisfies JWT,
          maxAge: 60,
        });
        return nexaut_header;
      },
    }),
  ],
  transformer: SuperJSON,
});
