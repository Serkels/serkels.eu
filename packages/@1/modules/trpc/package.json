{
  "name": "@1.modules/trpc",
  "version": "0.1.0",
  "private": true,
  "sideEffects": false,
  "type": "module",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "scripts": {
    "dev": "tsc --build --watch",
    "type-check": "tsc --build"
  },
  "keywords": [],
  "dependencies": {
    "@1.infra/database": "workspace:*",
    "@1.infra/email": "workspace:*",
    "@1.modules/profile.domain": "workspace:*",
    "@douglasduteil/nextauth...trpc.prisma": "workspace:*",
    "@trpc/server": "10.43.6",
    "superjson": "2.2.1",
    "zod": "3.22.4"
  },
  "devDependencies": {
    "@1.config/typescript": "workspace:*",
    "typescript": "5.2.2"
  }
}
