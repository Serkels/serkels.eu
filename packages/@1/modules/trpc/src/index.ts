//

import type { Profile } from "@1.modules/profile.domain";
import { create_nextauth_header as any_create_nextauth_header } from "@douglasduteil/nextauth...trpc.prisma/jwt";
import type { ProcedureBuilder, ProcedureParams } from "@trpc/server";

//

export type { Context } from "./context";
export const create_nextauth_header = any_create_nextauth_header<{
  profile: Pick<Profile, "id" | "role">;
}>;

export {
  createCallerFactory,
  mergeRouters,
  middleware,
  procedure,
  router,
} from "./trpc";

export {
  maybe_next_auth_procedure,
  next_auth_input_token,
  next_auth_procedure,
} from "./guards";

export type inferProcedureInput<TQuery extends ProcedureBuilder<any>> =
  TQuery extends ProcedureBuilder<infer TParams>
    ? TParams extends ProcedureParams
      ? { ctx: TParams["_ctx_out"]; input: TParams["_input_out"] }
      : never
    : never;

//

export const with_previous_cursor =
  <TData>(limit: number, data: TData[]) =>
  <TCursor>(predicate: (data: TData) => TCursor) => {
    let prevCursor: TCursor | undefined = undefined;
    if (data.length > limit) {
      const prev_item = data.pop()!;
      prevCursor = predicate(prev_item);
    }

    return { data, prevCursor };
  };

export const with_next_cursor =
  <TData>(limit: number, data: TData[]) =>
  <TCursor>(predicate: (data: TData) => TCursor) => {
    let next_cursor: TCursor | undefined = undefined;
    if (data.length > limit) {
      const next_item = data.pop()!;
      next_cursor = predicate(next_item);
    }

    return { data, next_cursor };
  };
