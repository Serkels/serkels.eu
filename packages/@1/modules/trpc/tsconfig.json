{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@1.config/typescript/base.json",
  "compilerOptions": {}
}
