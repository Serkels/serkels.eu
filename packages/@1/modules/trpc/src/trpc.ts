//

import type { Profile } from "@1.modules/profile.domain";
import { NEXTAUTH_TRPCENV } from "@douglasduteil/nextauth...trpc.prisma/config";
import { decode } from "@douglasduteil/nextauth...trpc.prisma/jwt";
import { verify_next_auth_token } from "@douglasduteil/nextauth...trpc.prisma/trpc";
import { TRPCError, initTRPC } from "@trpc/server";
import SuperJSON from "superjson";
import { ZodError, z } from "zod";
import type { Context } from "./context";

//
export const { router, middleware, mergeRouters, procedure } = initTRPC
  .context<Context>()
  .create({
    transformer: SuperJSON,
    errorFormatter({ shape, error }) {
      return {
        ...shape,
        data: {
          ...shape.data,
          zodError:
            error.code === "BAD_REQUEST" && error.cause instanceof ZodError
              ? error.cause.flatten()
              : null,
        },
      };
    },
  });

// export type Procedure = inferProcedureBuilderResolverOptions<typeof procedure>;
export const next_auth_procedure = procedure.use(
  verify_next_auth_token<{ profile: Profile }>(
    NEXTAUTH_TRPCENV.NEXTAUTH_SECRET,
  ),
);

// export type Next_Auth_Procedure = inferProcedureBuilderResolverOptions<
//   typeof next_auth_procedure
// >;

export const next_auth_input_token = procedure
  .input(z.object({ token: z.string() }))
  .use(async ({ input, ctx, next }) => {
    try {
      const payload = (await decode({
        token: input.token,
        secret: NEXTAUTH_TRPCENV.NEXTAUTH_SECRET,
      })) as { profile: Profile };

      if (payload) {
        return next({ ctx: { ...ctx, payload } });
      }

      throw new TRPCError({
        code: "UNAUTHORIZED",
        cause: new Error("No payload"),
      });
    } catch (cause) {
      throw new TRPCError({ code: "PARSE_ERROR", cause });
    }
  });
