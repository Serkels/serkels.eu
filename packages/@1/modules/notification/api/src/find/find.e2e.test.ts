//

import {
  create_douglas_student,
  create_greatings_message,
  create_joedart_student,
} from "@1.infra/database/seeding";
import prisma, { empty_database, migrate } from "@1.infra/database/testing";
import { createCallerFactory } from "@1.modules/trpc";
import { douglas_student_session } from "@1.modules/trpc/testing";
import {
  afterEach,
  beforeAll,
  beforeEach,
  expect,
  setSystemTime,
  test,
} from "bun:test";
import find_api_router from "./find";

//

beforeAll(empty_database);
beforeAll(migrate);
beforeAll(() => {
  setSystemTime(new Date("2011-11-11"));
});

//

test("should return latest notifications", async () => {
  const caller = createCallerFactory(find_api_router);
  const trpc = caller({
    auth: () => douglas_student_session,
    prisma,
  } as any);
  const exchange = await trpc.find({});
  expect(exchange).toMatchSnapshot();
});

//

beforeEach(async function seed() {
  await create_douglas_student(prisma);
  await create_joedart_student(prisma);
  await create_greatings_message(
    prisma,
    "douglas_profile_id",
    "joedart_profile_id",
  );
});

afterEach(async function empty_delete_entries() {
  await prisma.category.deleteMany({});
  await prisma.notification.deleteMany({});
  await prisma.user.deleteMany({});
});
