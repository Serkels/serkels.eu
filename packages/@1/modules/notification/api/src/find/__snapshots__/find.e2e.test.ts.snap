// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should return latest notifications 1`] = `
{
  "data": [
    {
      "created_at": 2011-11-11T00:00:00.000Z,
      "exchange_message": null,
      "forum_message": null,
      "id": "greatings_notification_id",
      "inbox_message": {
        "message": {
          "author": {
            "id": "joedart_profile_id",
            "image": "https://picsum.photos/200/300",
            "name": "Joe Dart",
          },
          "thread_id": "greatings_thread_id",
        },
      },
      "owner_id": "douglas_profile_id",
      "profile_added": null,
      "read_at": null,
      "type": "INBOX_NEW_MESSAGE",
    },
  ],
  "prevCursor": undefined,
}
`;
